action/candidate_endorsement.go

```diff
@@ -146,16 +252,44 @@ func NewCandidateEndorsement(nonce, gasLimit uint64, gasPrice *big.Int, bucketIn
 	}
 }
 
+// NewCandidateEndorsement returns a CandidateEndorsement action
+func NewCandidateEndorsement(nonce, gasLimit uint64, gasPrice *big.Int, bucketIndex uint64, op CandidateEndorsementOp) *CandidateEndorsement {
+	return &CandidateEndorsement{
+		AbstractAction: AbstractAction{
+			version:  version.ProtocolVersion,
+			nonce:    nonce,
+			gasLimit: gasLimit,
+			gasPrice: gasPrice,
+		},
+		bucketIndex: bucketIndex,
+		op:          op,
+	}
+}
+
 // NewCandidateEndorsementFromABIBinary parses the smart contract input and creates an action
 func NewCandidateEndorsementFromABIBinary(data []byte) (*CandidateEndorsement, error) {
+	if len(data) <= 4 {
+		return nil, errDecodeFailure
+	}
+	switch {
+	case bytes.Equal(candidateEndorsementMethod.ID, data[:4]):
+		return newCandidateEndorsementFromABIBinaryLegacy(data)
+	case bytes.Equal(candidateEndorseMethod.ID, data[:4]):
+		return newCandidateEndorseFromABIBinary(data)
+	case bytes.Equal(candidateIntentToRevokeEndorsementMethod.ID, data[:4]):
+		return newCandidateIntentToRevokeEndorsementFromABIBinary(data)
+	case bytes.Equal(candidateRevokeEndorsementMethod.ID, data[:4]):
+		return newCandidateRevokeEndorsementFromABIBinary(data)
+	default:
+		return nil, errDecodeFailure
```

Comment:
similar to `encodeABI`, just set the method and op in switch statement, we don't need 4 funcs
```
switch {
	case xxx:
	    method = candidateEndorseMethod
	    op = OpEndorse
	case yyy:
	    method = candidateUnendorseMethod
	    op = OpUnendorse
 }
 bucketID, ok := paramsMap["bucketIndex"].(uint64)
	if !ok {
		return nil, errDecodeFailure
	}
	cr.bucketIndex = bucketID
	cr.op = op
	return &cr, nil
```